source/board/microbitv2/nrf52820/power.c

```diff
@@ -19,19 +19,190 @@
  * limitations under the License.
  */
 
+#include "nrf.h"
+
+#include "gpio.h"
+#include "gpio_extra.h"
+#include "main_interface.h"
+#include "pwr_mon.h"
+#include "IO_Config.h"
+#include "DAP_config.h"
+#include "uart.h"
+#include "i2c.h"
+#include "rl_usb.h"
+
 #include "power.h"
 
+
+static void power_before(bool systemoff);
+static void power_after();
+static void power_systemoff();
+static void power_wfi();
+
+void POWER_CLOCK_IRQHandler(void);
+void GPIOTE_IRQHandler(void);
+
+
+extern volatile uint8_t wake_from_reset;
+extern volatile uint8_t wake_from_usb;
+extern volatile bool usb_pc_connected;
+extern main_usb_connect_t usb_state;
+extern power_source_t power_source;
+
+
+uint8_t  power_in_wfi;
+uint8_t  power_gpiote_enabled;
+uint32_t power_gpiote_intenset;
+
+
 void power_init()
 {
-    // TODO
+    power_in_wfi = 0;
+
+    gpio_cfg_input(GPIO_REG(RESET_BUTTON), GPIO_IDX(RESET_BUTTON), RESET_BUTTON_PULL);
+
+    // Enable NRF_POWER interrupt for USB removed/detected
+    NRF_POWER->INTENSET = POWER_INTENSET_USBREMOVED_Msk | POWER_INTENSET_USBDETECTED_Msk;
+    NVIC_EnableIRQ(POWER_CLOCK_IRQn);
```

Comment:
Is this what wakes ups the interface on USB cable insertion? Is there no checking of the WAKE_ON_EDGE signal?

Reply:
That's the way I found. How to check the WAKE_ON_EDGE signal?

Reply:
WAKE_ON_EDGE is now called VBUS_ABSENT in the latest schematics. If the signal goes from high to low it means USB power has been inserted and that pin can be set up as an interrupt to wake up the IF MCU. Does the KL27 not do the same?

Reply:
Oh interesting, my only concern would be if the BAT60J (D51) from `VCOMBINED` could leak some voltage from the battery into the USB VBus.
`VBUS_ABSENT` (previously named `WAKE_ON_EDGE`) has worked well on KL27, so I'd probably try that first and keep this USBDETECTED in mind if we need to revisit.

Do you think there would be any advantages to use this instead of `VBUS_ABSENT`?

> What pin number is VBUS_ABSENT? 

nRF52820 -> P0.7
nRF52833 -> P0.11
It's identified as `VBUS_ABSENT` in the schematics.

Reply:
This is all clearer now I have seen the schematics.
I have switched to using VBUS_ABSENT for USB detection.